# <center> Final Exam

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# 1

In [2]:
caravan=pd.read_csv('Caravan.csv')
caravan.head()

MOSTYPE  MAANTHUI  MGEMOMV  MGEMLEEF  MOSHOOFD  MGODRK  MGODPR  MGODOV  \
0       33         1        3         2         8       0       5       1   
1       37         1        2         2         8       1       4       1   
2       37         1        2         2         8       0       4       2   
3        9         1        3         3         3       2       3       2   
4       40         1        4         2        10       1       4       1   

   MGODGE  MRELGE    ...     APERSONG  AGEZONG  AWAOREG  ABRAND  AZEILPL  \
0       3       7    ...            0        0        0       1        0   
1       4       6    ...            0        0        0       1        0   
2       4       3    ...            0        0        0       1        0   
3       4       5    ...            0        0        0       1        0   
4       4       7    ...            0        0        0       1        0   

   APLEZIER  AFIETS  AINBOED  ABYSTAND  Purchase  
0         0       0        0         0        No  
1         0       0        0         0        No  
2         0       0        0         0        No  
3         0       0        0         0        No  
4         0       0        0         0        No  

[5 rows x 86 columns]

In [11]:
caravan.Purchase=caravan.Purchase.replace('No',0)
caravan.Purchase=caravan.Purchase.replace('Yes',1)
caravan['Purchase'] = caravan['Purchase'].astype(str)

In [12]:
X=caravan.drop(columns=['PVRAAUT','AVRAAUT','Purchase'],axis=1)
y=caravan['Purchase']

In [13]:
X_test=X[:1000]
X_train=X[1000:]
y_test=y[:1000]
y_train=y[1000:]

## 1 a)

In [6]:
from sklearn.ensemble import RandomForestClassifier

In [7]:
forest=RandomForestClassifier(max_features=29,n_estimators=500,random_state=0)
forest.fit(X_train,y_train)

RandomForestClassifier(max_features=29, n_estimators=500, random_state=0)

In [8]:
forest_pred=pd.DataFrame()
forest_pred['y_test']=y_test
forest_pred['y_pred']=forest.predict(X_test)

In [9]:
pd.crosstab(forest_pred['y_pred'],forest_pred['y_test'])

y_test    0   1
y_pred         
0       915  52
1        26   7

In [10]:
7/(26+7)

0.21212121212121213

The test set precision for the random forest model is 0.2121.

## 1 b)

In [11]:
from sklearn.ensemble import GradientBoostingClassifier

In [12]:
boosting=GradientBoostingClassifier(n_estimators=1000,learning_rate=0.01,\
                                    max_depth=4,random_state=0)
boosting.fit(X_train,y_train)

GradientBoostingClassifier(learning_rate=0.01, max_depth=4, n_estimators=1000,
                           random_state=0)

In [13]:
boosting_pred=pd.DataFrame()
boosting_pred['y_test']=y_test
boosting_pred['y_pred']=boosting.predict(X_test)

In [14]:
pd.crosstab(boosting_pred['y_pred'],boosting_pred['y_test'])

y_test    0   1
y_pred         
0       937  57
1         4   2

In [15]:
2/(4+2)

0.3333333333333333

The test set precision for the boosting model is 0.3333.

## 1 c)

In [6]:
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier

In [7]:
scaler=StandardScaler()
scaler.fit(X_train)
X_train_scaled=scaler.transform(X_train)
X_test_scaled=scaler.transform(X_test)

In [8]:
mlp=MLPClassifier(solver='lbfgs',random_state=0,activation='tanh',\
                  hidden_layer_sizes=[400,400,400],max_iter=10000)
mlp.fit(X_train_scaled,y_train)

MLPClassifier(activation='tanh', hidden_layer_sizes=[400, 400, 400],
              max_iter=10000, random_state=0, solver='lbfgs')

In [9]:
mlp_pred=pd.DataFrame()
mlp_pred['y_test']=y_test
mlp_pred['y_pred']=mlp.predict(X_test_scaled)

In [10]:
pd.crosstab(mlp_pred['y_pred'],mlp_pred['y_test'])

y_test    0   1
y_pred         
0       884  48
1        57  11

In [21]:
11/(57+11)

0.16176470588235295

The test set precision for the mlp model is 0.1618.

## 1 d)

In [22]:
from sklearn.linear_model import LogisticRegression

In [23]:
lr=LogisticRegression(solver='lbfgs',max_iter=10000)
lr.fit(X_train_scaled,y_train)

LogisticRegression(max_iter=10000)

In [24]:
lr_pred=pd.DataFrame()
lr_pred['y_test']=y_test
lr_pred['y_pred']=lr.predict(X_test_scaled)

In [25]:
pd.crosstab(lr_pred['y_pred'],lr_pred['y_test'])

y_test    0   1
y_pred         
0       934  59
1         7   0

In [26]:
0/(7+0)

0.0

The test set precision for the logistic regression model is 0.

## 1 e)

In [27]:
lr_pred['y_proba']=lr.predict_proba(X_test_scaled)[:,1]

In [28]:
lr_pred.loc[lr_pred['y_proba']>0.25,'y_pred2']=1
lr_pred.loc[lr_pred['y_proba']<=0.25,'y_pred2']=0

In [29]:
pd.crosstab(lr_pred['y_pred2'],lr_pred['y_test'])

y_test     0   1
y_pred2         
0.0      917  48
1.0       24  11

In [30]:
11/(24+11)

0.3142857142857143

The test set precision for the logistic regression model with adjusted cut-off probability is 0.3143.

# 2

In [31]:
from sklearn.datasets import fetch_lfw_people
faces=fetch_lfw_people(min_faces_per_person=60)

In [32]:
print(faces.images.shape,faces.data.shape,faces.target.shape,faces.target_names.shape)

(1348, 62, 47) (1348, 2914) (1348,) (8,)


In [33]:
X=faces.data
y=faces.target

In [36]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.75,random_state=42,
                                              stratify=y)

## a)

In [37]:
from sklearn.svm import SVC
from sklearn.pipeline import make_pipeline

In [38]:
svc=SVC(kernel='rbf',gamma=0.01)
model1=make_pipeline(svc)
model1.fit(X_train,y_train)

Pipeline(steps=[('svc', SVC(gamma=0.01))])

In [39]:
model1.score(X_test,y_test)

0.39465875370919884

The test accuracy rate of model 1 is 0.3947.

## b)

In [40]:
from sklearn.decomposition import PCA

In [41]:
pca=PCA(n_components=150,svd_solver='randomized',whiten=True,random_state=42)
svc1=SVC(kernel='rbf',gamma=0.01)
model2=make_pipeline(pca,svc1)
model2.fit(X_train,y_train)

Pipeline(steps=[('pca',
                 PCA(n_components=150, random_state=42, svd_solver='randomized',
                     whiten=True)),
                ('svc', SVC(gamma=0.01))])

In [42]:
model2.score(X_test,y_test)

0.6973293768545994

The test accuracy rate of model 2 is 0.6973.

## c)

In [43]:
svc2=SVC(kernel='rbf',gamma=0.01,class_weight='balanced')
model3=make_pipeline(pca,svc2)
model3.fit(X_train,y_train)

Pipeline(steps=[('pca',
                 PCA(n_components=150, random_state=42, svd_solver='randomized',
                     whiten=True)),
                ('svc', SVC(class_weight='balanced', gamma=0.01))])

In [44]:
model3.score(X_test,y_test)

0.744807121661721

The test accuracy rate of model 3 is 0.7448.

## d)

In [45]:
from sklearn.model_selection import GridSearchCV

In [46]:
svc3=SVC(kernel='rbf',class_weight='balanced')
model4=make_pipeline(pca,svc3)
param_grid={'svc__C':[1,5,10],\
           'svc__gamma':[0.0001,0.0005,0.001,0.005]}
grid=GridSearchCV(model4,param_grid,cv=5)
grid.fit(X_train,y_train)
model4=grid.best_estimator_

In [47]:
model4.score(X_test,y_test)

0.8486646884272997

The test accuracy rate of model 4 is 0.8487.

## e)

We can see that model 4 is the best model.

In [48]:
from sklearn.metrics import confusion_matrix
y_pred=model4.predict(X_test)
matrix=pd.DataFrame(confusion_matrix(y_test,y_pred),index=model4.classes_,\
                    columns=model4.classes_)
matrix.index=faces.target_names
matrix.columns=faces.target_names
matrix

Ariel Sharon  Colin Powell  Donald Rumsfeld  George W Bush  \
Ariel Sharon                 10             4                0              5   
Colin Powell                  1            53                0              4   
Donald Rumsfeld               0             2               23              4   
George W Bush                 0             5                0            128   
Gerhard Schroeder             0             0                0              2   
Hugo Chavez                   0             3                0              3   
Junichiro Koizumi             0             0                1              2   
Tony Blair                    0             0                1              6   

                   Gerhard Schroeder  Hugo Chavez  Junichiro Koizumi  \
Ariel Sharon                       0            0                  0   
Colin Powell                       0            0                  0   
Donald Rumsfeld                    0            1                  0   
George W Bush                      0            0                  0   
Gerhard Schroeder                 24            0                  0   
Hugo Chavez                        1           10                  0   
Junichiro Koizumi                  0            0                 10   
Tony Blair                         1            0                  0   

                   Tony Blair  
Ariel Sharon                0  
Colin Powell                1  
Donald Rumsfeld             0  
George W Bush               0  
Gerhard Schroeder           1  
Hugo Chavez                 1  
Junichiro Koizumi           2  
Tony Blair                 28

In [49]:
matrix_normalized = matrix / matrix.sum(axis=1)
matrix_array=matrix_normalized.values
error_rate=1-np.diag(matrix_array)
pred_diff=pd.DataFrame(error_rate,index=faces.target_names,columns=['error_rate'])
pred_diff

error_rate
Ariel Sharon         0.473684
Colin Powell         0.101695
Donald Rumsfeld      0.233333
George W Bush        0.037594
Gerhard Schroeder    0.111111
Hugo Chavez          0.444444
Junichiro Koizumi    0.333333
Tony Blair           0.222222

In [50]:
pred_diff[pred_diff.error_rate==pred_diff.error_rate.max()]

error_rate
Ariel Sharon    0.473684

"Ariel Sharon" has the largest error rate: 0.4737.